[GTK] Build fix for Ubuntu Stable after 274201@main

dpino opened the pull request:
#### ef5415e49bdf5f8d84f66b6946d698a4bc298188
<pre>
[GTK] Build fix for Ubuntu Stable after 274201@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=227528">https://bugs.webkit.org/show_bug.cgi?id=227528</a>

Reviewed by NOBODY (OOPS!).

Method &apos;get_first_accessible_child&apos; is only available since GTK 4.10.
Ubuntu Stable (Ubuntu 22.04) features GTK 4.6.9.

Guard code accessing &apos;get_first_accessible_child&apos; under
&apos;ifdef GTK_ACCESSIBILITY_ATSPI&apos;, since other parts of the code also
depend on this guard.

* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseAccessibleInterfaceInit):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef5415e49bdf5f8d84f66b6946d698a4bc298188

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38548 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40873 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41083 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34225 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/20343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14822 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32407 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39121 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14729 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33535 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12795 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12778 "Unexpected infrastructure issue, retrying build") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42359 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35013 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38606 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13388 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11065 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36812 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14988 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13855 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14462 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -419,6 +420,14 @@ webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface* iface)
+static void
+webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface* iface)
+{
+    UNUSED_PARAM(iface);
```
You can just omit the parameter name:

```suggestion
webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*)
{
```

(Source/WebKit is built with -Wno-unused-parameter, so even that is not strictly necessary, but best to write the code as if it is used.)

mcatanzaro reviewed (commented):
This is not really ideal because it leaves `webView->priv->socketAccessible` around but unused. It would be clearer to try to disable all of the code that depends on GTK 4.10.

dpino committed:
[GTK] Build fix for Ubuntu Stable after 274201@main
https://bugs.webkit.org/show_bug.cgi?id=227528

Reviewed by NOBODY (OOPS!).

Method 'get_first_accessible_child' is only available since GTK 4.10.
Ubuntu Stable (Ubuntu 22.04) features GTK 4.6.9.

Guard code accessing 'get_first_accessible_child' under
'ifdef GTK_ACCESSIBILITY_ATSPI', since other parts of the code also
depend on this guard.

* Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
(webkitWebViewBaseAccessibleInterfaceInit):

GeorgesStavracas commented on Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:
```diff
@@ -417,7 +417,9 @@ static GtkAccessible* webkitWebViewBaseAccessibleGetFirstAccessibleChild(GtkAcce
 static void
 webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface* iface)
 {
+#ifdef GTK_ACCESSIBILITY_ATSPI
     iface->get_first_accessible_child = webkitWebViewBaseAccessibleGetFirstAccessibleChild;
+#endif
```
I think something like what Carlos mentioned would work, without wornings. Basically:

```c
#if USE(GTK4)

#ifdef GTK_ACCESSIBILITY_ATSPI
static void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*);
WEBKIT_DEFINE_TYPE_WITH_CODE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET,
    G_IMPLEMENT_INTERFACE(GTK_TYPE_ACCESSIBLE, webkitWebViewBaseAccessibleInterfaceInit))
#else
WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET)
#endif

[... GTK3 defines ...]

#endif
```

Then this section would look like:

```c
#if USE(GTK4) && defined(GTK_ACCESSIBILITY_ATSPI)
static GtkAccessible* webkitWebViewBaseAccessibleGetFirstAccessibleChild(GtkAccessible* accessible)
{
    WebKitWebViewBase* webView = WEBKIT_WEB_VIEW_BASE(accessible);

    if (webView->priv->socketAccessible)
        return GTK_ACCESSIBLE(g_object_ref(webView->priv->socketAccessible.get()));

    if (auto* widget = gtk_widget_get_first_child(GTK_WIDGET(webView)))
        return GTK_ACCESSIBLE(g_object_ref(widget));

    return nullptr;
}

static void
webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface* iface)
{
    iface->get_first_accessible_child = webkitWebViewBaseAccessibleGetFirstAccessibleChild;
}
#endif

dpino replied:
Applying this patch I get the following build error:

```
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:387:13: warning: ‘void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*)’ declared ‘static’ but never defined [-Wunused-function]
  387 | static void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*);
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Then, if I try to guard the declaration of webkitWebViewBaseAccessibleInterfaceInit with `#if USE(GTK4) && defined(GTK_ACCESSIBILITY_ATSPI)`, I get this other error:

```
In file included from /home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:93:
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp: In function ‘GType webkit_web_view_base_get_type_once()’:
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:392:61: error: ‘GTK_TYPE_CONTAINER’ was not declared in this scope; did you mean ‘GTK_TYPE_CONSTRAINT’?
  392 | WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_CONTAINER)
	  |                                                             ^~~~~~~~~~~~~~~~~~
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/WebKitBuild/GTK/Release/WTF/Headers/wtf/glib/WTFGType.h:112:13: note: in definition of macro ‘_WEBKIT_DEFINE_TYPE_EXTENDED_BEGIN’
  112 |             TYPE_PARENT, \
	  |             ^~~~~~~~~~~
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/WebKitBuild/GTK/Release/WTF/Headers/wtf/glib/WTFGType.h:42:62: note: in expansion of macro ‘_WEBKIT_DEFINE_TYPE_EXTENDED’
   42 | #define WEBKIT_DEFINE_TYPE(TypeName, type_name, TYPE_PARENT) _WEBKIT_DEFINE_TYPE_EXTENDED(TypeName, type_name, TYPE_PARENT, 0, { })
	  |                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/buildbot/worker/GTK-Linux-64-bit-Release-Ubuntu-LTS-Build/build/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp:392:1: note: in expansion of macro ‘WEBKIT_DEFINE_TYPE’
  392 | WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_CONTAINER)
	  | ^~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

GeorgesStavracas replied:
This diff seems to work for me:

```diff
diff --git a/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp b/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
index bbe3b935d7..be5781d7e9 100644
--- a/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
+++ b/Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp
@@ -384,10 +384,16 @@ struct _WebKitWebViewBasePrivate {
  */
 
 #if USE(GTK4)
+
+#ifdef GTK_ACCESSIBILITY_ATSPI
 static void webkitWebViewBaseAccessibleInterfaceInit(GtkAccessibleInterface*);
 
 WEBKIT_DEFINE_TYPE_WITH_CODE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET,
     G_IMPLEMENT_INTERFACE(GTK_TYPE_ACCESSIBLE, webkitWebViewBaseAccessibleInterfaceInit))
+#else
+WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_WIDGET)
+#endif // GTK_ACCESSIBILITY_ATSPI
+
 #else
 WEBKIT_DEFINE_TYPE(WebKitWebViewBase, webkit_web_view_base, GTK_TYPE_CONTAINER)
 #endif
@@ -398,15 +404,13 @@ static void webkitWebViewBaseDidExitFullScreen(WebKitWebViewBase*);
 static void webkitWebViewBaseRequestExitFullScreen(WebKitWebViewBase*);
 #endif
 
-#if USE(GTK4)
+#if USE(GTK4) && defined(GTK_ACCESSIBILITY_ATSPI)
 static GtkAccessible* webkitWebViewBaseAccessibleGetFirstAccessibleChild(GtkAccessible* accessible)
 {
     WebKitWebViewBase* webView = WEBKIT_WEB_VIEW_BASE(accessible);
 
-#ifdef GTK_ACCESSIBILITY_ATSPI
     if (webView->priv->socketAccessible)
         return GTK_ACCESSIBLE(g_object_ref(webView->priv->socketAccessible.get()));
-#endif
 
     if (auto* widget = gtk_widget_get_first_child(GTK_WIDGET(webView)))
         return GTK_ACCESSIBLE(g_object_ref(widget));
```

dpino replied:
@GeorgesStavracas Thank you, this works. Since you authorized the patch I think it's better if you push the patch in a new PR (mark the PR as Unreviewed) and I will land it.

GeorgesStavracas replied:
Sure: https://github.com/WebKit/WebKit/pull/24066
